src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,447 @@
+
+## Notations
+
+ * `E⟮K⟯`: the group of `K`-rational points on an elliptic curve `E`.
+
+## Implementation notes
+
+The proofs of `EllipticCurve.point.weierstrass_dbl'`, `EllipticCurve.point.weierstrass_add'`, and
+`EllipticCurve.point.add_assoc` (TODO) are slow due to the sheer size of the polynomials involved,
+but they can be hastened considerably by breaking the computations into multiple steps if necessary.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, group law
+-/
+
+namespace EllipticCurve
+
+open polynomial
+
+open_locale polynomial
+
+universes u v
+
+section basic
+
+/-!
+### The `K`-rational points `E⟮K⟯`
+
+Use `0` instead of `EllipticCurve.point.zero`.
+-/
+
+variables {F : Type u} [comm_ring F] (E : EllipticCurve F) (K : Type v) [comm_ring K] [algebra F K]
+
+section weierstrass
+
+/-- The Weierstrass polynomial $w_E(X, Y) = Y^2 + a_1XY + a_3Y - (X^3 + a_2X^2 + a_4X + a_6)$. -/
+noncomputable def weierstrass_polynomial : K[X][X] :=
+X ^ 2 + C (C ↑E.a₁ * X) * X + C (C ↑E.a₃) * X - C (X ^ 3 + C ↑E.a₂ * X ^ 2 + C ↑E.a₄ * X + C ↑E.a₆)
+
+variables {K}
+
+/-- The proposition that an affine point $(x, y)$ lies in `E`, that is $w_E(x, y) = 0$. -/
+def weierstrass_equation (x y : K) : Prop := eval x (eval (C y) $ E.weierstrass_polynomial K) = 0
+
+@[simp] lemma eval_weierstrass_polynomial (x y : K) :
+  eval x (eval (C y) $ E.weierstrass_polynomial K)
+    = y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y - (x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆) :=
+by simp only [weierstrass_polynomial, eval_sub, eval_add, eval_mul, eval_pow, eval_C, eval_X]
+
+lemma weierstrass_equation_iff' (x y : K) :
+  E.weierstrass_equation x y
+    ↔ y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y - (x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆) = 0 :=
+by rw [weierstrass_equation, eval_weierstrass_polynomial]
+
+@[simp] lemma weierstrass_equation_iff (x y : K) :
+  E.weierstrass_equation x y
+    ↔ y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y = x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆ :=
+by rw [weierstrass_equation_iff', sub_eq_zero]
+
+end weierstrass
+
+variables (K)
+
+/-- The group of `K`-rational points `E⟮K⟯` on an elliptic curve `E` over `F`. This consists of the
+unique point at infinity `EllipticCurve.point.zero`and the affine points `EllipticCurve.point.some`
+satisfying the Weierstrass equation $y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6$ of `E`. -/
+inductive point
+| zero
+| some (x y : K) (h : E.weierstrass_equation x y)
+
+localized "notation E⟮K⟯ := point E K" in EllipticCurve
+
+variables {E K}
+
+namespace point
+
+instance : inhabited E⟮K⟯ := ⟨zero⟩
+
+section zero
+
+instance : has_zero E⟮K⟯ := ⟨zero⟩
+
+@[simp] lemma zero_def : zero = (0 : E⟮K⟯) := rfl
+
+end zero
+
+section negation
+
+variables {x y : K} (h : E.weierstrass_equation x y)
+
+include h
+
+/-- The $y$-coordinate of the negation of an affine point. -/
+@[nolint unused_arguments] def neg_y : K := -y - ↑E.a₁ * x - ↑E.a₃
```

Comment:
The `@[nolint unused_arguments]` should be harmless. In any case, are you saying you want to define some kind of `neg : K x K -> K x K` on the affine plane? How would it be defined if does not refer to `E` at all?

Reply:
The "conjugate" `-Y - a_1 X - a_3` of the polynomial `Y` is used in the [proof here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/thoughts.20on.20elliptic.20curves/near/311534577), so we certainly want to define the polynomial `-X - E.a₁ * C X - E.a₃` in `K[X][X]`. And yes it needs to refer to `E` (but only two of its coefficients); that's why I said you need to make `E` explicit. Of course, `neg_y` is the evaluation of this polynomial, and I think it could be convenient to use properties of `neg_y` for general `(x, y)` not necessarily on the curve when we solve the equations and apply nullstellensatz there; we'll know more as we work out the details. By the way, do you have a better way to show `to_class` respect addition than my outline there?

Reply:
I've defined `neg_polynomial` as an element of `K[X][X]`, but I think we should keep `neg_y` with implicit variables as it is right now, with a lemma that says evaluating `neg_polynomial` gives `neg_y h`. The reason is partly for definitional equality (otherwise `weierstrass_equation_dbl`, etc complains) and partly to avoid adding `noncomputable` to `neg`, etc (I guess this is not too important), and I think the lemma should give everything we need. What do you think?

I also added `dbl_polynomial` for the numerator of the slope by defining it as an explicit polynomial, but it should really be the x-derivative of the Weierstrass polynomial, yet I seem to struggle to find an existing way to map across Weierstrass polynomial over `polynomial.derivative`. Once we do have this, do you still want the function `dbl`?

I will push my code once I make a PR for the lemmas needed in `base_change` in `EllipticCurve.lean`. As for `to_class`, I haven't had time to think about it at all, so my plan is to follow your outline.

Reply:
I changed the `neg_Y` etc to take in `E x y` rather than `h` - you're right that it made some things a lot more convenient.